src/inventory/beacon_inventory.rs

```diff
@@ -0,0 +1,481 @@
+                    self.info.effect1 = this_btn_effect;
+                    if self.info.power_level == 4 {
+                        get_button(7, basic_elements).texture_coords = BUTTON_ACTIVE;
+                    }
+                    get_button(15, basic_elements).texture = current_btn_image;
+                    get_button(15, basic_elements).colour.3 = 255;
+                    get_button(7, basic_elements).colour.3 = 255;
+
+                    //reset previously pressed buttons
+                    for j in 0..n.min(5) {
+                        if j != i {
+                            get_button(j + 1, basic_elements).texture_coords = BUTTON_ACTIVE;
+                        }
+                    }
+                }
+
+                // choose effect level 2  or regen
+                if self.info.power_level == 4
+                    && get_button_read(i + 1, &basic_elements).texture_coords == BUTTON_PRESSED
+                {
+                    if get_button_read(6, &basic_elements).texture_coords == BUTTON_PRESSED
+                        && self.info.effect2 != Some(Effect::Regeneration)
+                    {
+                        get_button(7, basic_elements).texture_coords = BUTTON_ACTIVE;
+                        self.info.effect2 = Some(Effect::Regeneration);
+                    } else if get_button_read(7, &basic_elements).texture_coords == BUTTON_PRESSED
+                        && self.info.effect2 != this_btn_effect
+                    {
+                        get_button(6, basic_elements).texture_coords = BUTTON_ACTIVE;
+                        self.info.effect2 = this_btn_effect;
+                    }
+                }
+
+                // make the tick active
+                if let Some(item) = self.slots.get_item(0) {
+                    if item.material == Material::GoldIngot
+                        || item.material == Material::Diamond
+                        || item.material == Material::IronIngot
+                        || item.material == Material::Emerald
+                        || item.material == Material::NetheriteIngot
+                    {
+                        get_button(8, basic_elements).texture_coords = BUTTON_ACTIVE;
+                    } else {
+                        get_button(8, basic_elements).texture_coords = BUTTON_INACTIVE;
+                    }
+                } else {
+                    get_button(8, basic_elements).texture_coords = BUTTON_INACTIVE;
+                }
+            }
+        }
+    }
+
+    fn ty(&self) -> InventoryType {
+        InventoryType::Beacon
+    }
+}
+
+fn get_button(i: usize, basic_elements: &mut Vec<ui::ImageRef>) -> std::cell::RefMut<ui::Image> {
+    basic_elements.get_mut(i).unwrap().borrow_mut()
+}
+
+fn get_button_read<'a>(
+    i: usize,
+    basic_elements: &'a &'a mut Vec<ui::ImageRef>,
+) -> std::cell::Ref<'a, ui::Image> {
+    basic_elements.get(i).unwrap().borrow()
+}
```

Comment:
Rather than comparing textures from `get_button_read`, could you change this to return an enum value? Maybe return something like
```
enum ButtonState {
    Active,
    Pressed,
    Inactive,
    Focused,
}
```
I think this would make that code trying to detect state changes a bit more simple.

Rather than `get_button(...)`, you could use `set_button_state(ButtonState, ...)` to switch textures.

Reply:
ok, ive made a commit implementing this and also added a few more functions to make the state changes more readable.